builder: Retry RPM downloads for build bundles

jwakre opened the pull request:
Download failures can occur when downloading large sets of RPMs from a
server with a poor connection. This change adds retry functionality to
attempt RPM downloads multiple times.

Fixes: #531

Signed-off-by: John Akre <john.w.akre@intel.com>

otaviobp commented on builder/bundles.go:
```diff
@@ -447,6 +447,22 @@ func installBundleToFull(packagerCmd []string, buildVersionDir string, bundle *b
 			args = append(args, p)
 		}
+
+		// Retry RPM downloads to avoid timeout failures due to slow network
```
I prefer to have a function downloadRpms(packagerCmd, baseDir, max_retries) to encapsulate that.

jwakre replied:
sure

otaviobp commented on builder/bundles.go:
```diff
@@ -447,6 +447,22 @@ func installBundleToFull(packagerCmd []string, buildVersionDir string, bundle *b
+			if retries < 2 {
+				retries++
+				fmt.Println("Retrying after failed RPM download")
+				goto DOWNLOAD_RPMS
```
I don't like gotos with the only exception for going to the end of a function on error handling.
It's a lot easier to read if you have

    while (retry < max_retries) {
         ...
         if (err == nil) break; //Or return if you create another function
    }

jwakre replied:
I was going back-and-forth on this. I can switch it to a loop implementation.

rchiossi replied:
I'm with otavio here, specially considering that feels out of place compared to the rest of the code base.

jwakre commented:
It looks like there is a problem with the webserver hosting download content.

rchiossi commented on builder/bundles.go:
```diff
@@ -436,17 +436,43 @@ func genUpdateBundleSpecialFiles(chrootDir string, b *Builder) error {
+	for retries := 0; retries < maxRetries; retries++ {
+		_, err = helpers.RunCommandOutputEnv(args[0], args[1:], []string{"LC_ALL=en_US.UTF-8"})
+		if err == nil {
+			return err
```
Doesn't really matter here, but you could also call `break` to end the `for` loop. No need to change it though.

rchiossi commented on builder/bundles.go:
```diff
@@ -436,17 +436,43 @@ func genUpdateBundleSpecialFiles(chrootDir string, b *Builder) error {
+		}
+
+		// Retry RPM downloads to avoid timeout failures due to slow network
+		err = downloadRpms(packagerCmd, rpmList, baseDir, 3)
```
It would be nicer to have a constant defined with the number of retries instead of having a magic number on the function call. I also think it would be nice if this parameter could be set via command line, so users with troublesome connection could thinker with a better value.

jwakre replied:
Ok, I can add this.

otaviobp commented:
I was looking at DNF manual (https://dnf.readthedocs.io/en/latest/conf_ref.html) and the standard number of retries in dnf is 10. Are we overriding this or we are going to retry dnf command after dnf failed 10 times?

And if we are overriding this, can't we just set the retry in dnf instead of adding that code to swupd?

I think it makes sense to retry on errors, but it's weird to have a failure at first. I'd prefer to have this problem investigated instead of just implementing a retry strategy.

Another question. Have you tried to reproduce that using dnf directly?

jwakre commented:
> I was looking at DNF manual (https://dnf.readthedocs.io/en/latest/conf_ref.html) and the standard number of retries in dnf is 10. Are we overriding this or we are going to retry dnf command after dnf failed 10 times?

The DNF command is using the .yum-conf from the Mixer workspace. I played with the timeout and retries parameters in this file, but I couldn't get rid of the failure this way. Something feels strange about the default mirror, but I couldn't find the root cause.

> And if we are overriding this, can't we just set the retry in dnf instead of adding that code to swupd?

I think this is default of 10.

> I think it makes sense to retry on errors, but it's weird to have a failure at first. I'd prefer to have this problem investigated instead of just implementing a retry strategy.

This is fair, I was just trying to unblock users.

> Another question. Have you tried to reproduce that using dnf directly?

No, but I can try it.

jwakre commented on builder/bundles.go:
```diff
@@ -436,17 +436,51 @@ func genUpdateBundleSpecialFiles(chrootDir string, b *Builder) error {
+		}
+
+		fmt.Printf("RPM download attempt %d failed. Retrying up to %d times.\n", retries+1, maxRetries)
+		if err := clearDNFCache(packagerCmd); err != nil {
```
I added the clearDNFCache function to remove potentially corrupted data from the DNF cache.

rchiossi commented:
If I understand correctly this last changes are to address the issue with embargo. @mbelluzzo is there any way you can verify if this fix your issue before we merge?

jwakre commented on builder/bundles.go:
```diff
@@ -436,17 +436,51 @@ func genUpdateBundleSpecialFiles(chrootDir string, b *Builder) error {
 
-func installBundleToFull(packagerCmd []string, buildVersionDir string, bundle *bundle) error {
+func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int) error {
+	var downloadErr error
```
Renamed err -> downloadErr so that it does not get overwritten by err := clearDNFCache(

mbelluzzo commented:
I can try this PR during this week on my infra and see how it does. The error is not predictable, so an one-shot try wont be enough. I can provide an update on this later this week.

mbelluzzo commented:
I left this running for the entire week and besides a single error I had on the first attempt (which never occurred again) the retry saved me a handful of builds so I'm happy with that. Even that we haven't root caused the real issue, this works as a workaround.

rchiossi commented:
If that is the case, I'll include it in our next release. Thanks for testing it out @mbelluzzo .

rchiossi commented on builder/bundles.go:
```diff
@@ -436,17 +436,51 @@ func genUpdateBundleSpecialFiles(chrootDir string, b *Builder) error {
+func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int) error {
+	var downloadErr error
+
+	if maxRetries <= 0 {
```
I believe the name is a bit confusing here since `retry` imply trying it again, but from this code, If I don't want mixer to retry a download if it fails I need to set retry to 1 instead of 0 which can be misleading. I think it would be better to check if `maxRetries >= 0` and look from `0` to `1 + maxRetries` instead, since we always wanna do at least one loop.

jwakre replied:
Good catch! I'll update.

jwakre commented:
@rchiossi I changed the logic so that retries sets the number of download retries instead of download attempts. Also, I renamed the retries variable to attempts which is more accurate.

rchiossi commented on mixer/cmd/build.go:
```diff
@@ -585,6 +592,12 @@ func init() {
+	buildBundlesCmd.Flags().IntVar(&buildFlags.downloadRetries, "retries", retriesDefault, "Number of retry attempts to download RPMs")
+	buildFormatBumpCmd.Flags().IntVar(&buildFlags.downloadRetries, "retries", retriesDefault, "Number of retry attempts to download RPMs")
+	buildFormatOldCmd.Flags().IntVar(&buildFlags.downloadRetries, "retries", retriesDefault, "Number of retry attempts to download RPMs")
+	buildUpstreamFormatCmd.Flags().IntVar(&buildFlags.downloadRetries, "retries", retriesDefault, "Number of retry attempts to download RPMs")
```
Shouldn't it be a top level flag instead like bundle-workers?

jwakre replied:
There is already a bundle-workers flag that specifies the number of parallel workers when building bundles

rchiossi replied:
I know, I used it as an exemple of a top level flag.

rchiossi replied:
I'm suggesting that you replace all this code by:
`buildCmd.PersiustentFlags().IntVar(&buildFlags.downloadRetries, "retries", retriesDefault, "Number of retry attempts to download RPMs")`

jwakre replied:
Oh, I see what you mean now... I can do that. It just seems weird to me that the option would be exposed to ~3 commands that don't use it.

jwakre replied:
Updated the PR to make it a persistent flag
